include/isotp.h

```diff
@@ -0,0 +1,350 @@
+			isotp_tx_callback_t complete_cb, void *cb_arg,
+			s32_t timeout);
+
+/**
+ * @brief Send data with buffered context
+ *
+ * This function is similar to iso_tp_send_ctx_buf but the data is carried in
+ * a net_buf. net_buf_unref is called on the net_buf when sending is completed.
+ *
+ * @param can_dev     The CAN device to be used for sending and receiving.
+ * @param data        Data to be sent.
+ * @param len         Length of the data to be sent.
+ * @param rx_addr     Identifier for FC frames.
+ * @param tx_addr     Identifier for outgoing frames the receiver listens on.
+ * @param complete_cb Function called on completion or NULL.
+ * @param cb_arg      Argument passed to the complete callback.
+ * @param timeout     Timeout for buffer allocation.
+ *
+ * @retval ISO_TP_N_OK on success or any of ISO_TP_* on error
+ */
+int iso_tp_send_net_ctx_buf(struct device *can_dev,
+			    struct net_buf *data,
+			    const struct iso_tp_msg_id *tx_addr,
+			    const struct iso_tp_msg_id *rx_addr,
+			    isotp_tx_callback_t complete_cb, void *cb_arg,
+			    s32_t timeout);
+
+#endif /*CONFIG_CAN_ISOTP_ENABLE_CONTEXT_BUFFERS*/
+
+#if defined(CONFIG_CAN_ISOTP_USE_TX_BUF) && \
+	defined(CONFIG_CAN_ISOTP_ENABLE_CONTEXT_BUFFERS)
+/**
+ * @brief Send data with buffered context
+ *
+ * This function is similar to iso_tp_send but the context is automatically
+ * allocated from an internal pool and the data to be send is buffered in an
+ * internal net_buff.
+ *
+ * @param can_dev     The CAN device to be used for sending and receiving.
+ * @param data        Data to be sent.
+ * @param len         Length of the data to be sent.
+ * @param rx_addr     Identifier for FC frames.
+ * @param tx_addr     Identifier for outgoing frames the receiver listens on.
+ * @param complete_cb Function called on completion or NULL.
+ * @param cb_arg      Argument passed to the complete callback.
+ * @param timeout     Timeout for buffer allocation.
+ *
+ * @retval ISO_TP_N_OK on success or any of ISO_TP_* on error
+ */
+int iso_tp_send_buf(struct device *can_dev,
+		    const u8_t *data, size_t len,
+		    const struct iso_tp_msg_id *tx_addr,
+		    const struct iso_tp_msg_id *rx_addr,
+		    isotp_tx_callback_t complete_cb, void *cb_arg,
+		    s32_t timeout);
+#endif
+
+/** @cond INTERNAL_HIDDEN */
+
+struct iso_tp_callback {
+	isotp_tx_callback_t cb;
+	void *arg;
+};
+
+struct iso_tp_send_ctx {
+	int filter_id;
+	struct device *can_dev;
+	union {
+		struct net_buf *buf;
+		struct {
+			const u8_t *data;
+			size_t len;
+		};
+	};
+	u32_t error_nr;
+	u8_t wft;
+	u8_t bs;
+	struct iso_tp_fc_opts opts;
+	u8_t sn : 4;
+	u8_t is_net_buf  : 1;
+	u8_t is_ctx_slab : 1;
+	u8_t has_callback: 1;
+	u8_t state;
+	u8_t tx_backlog;
+	struct k_work work;
+	struct k_timer timeout;
+	union {
+		struct iso_tp_callback fin_cb;
+		struct k_sem fin_sem;
+	};
+	struct iso_tp_msg_id rx_addr;
+	struct iso_tp_msg_id tx_addr;
+};
+
+struct iso_tp_rcv_ctx {
```

Comment:
Could we rename this to `iso_tp_recv_ctx` which would make this the same length as the send struct name, and typically we have been calling receive as recv in other part of the code.

Reply:
Sure.